### PR TITLE
Add reproducibility percentage to daily build summary

### PR DESCRIPTION
**Pull request by adamfarley**
Where we have implemented testing for a reproducible build, this code ensures that the results of that testing are included in the daily build summary.

If we did try to run the tests, but failed for whatever reason, we provide a short summary of the relevant platforms and provide a lot more information in the "details" link to the jenkins job that generates the summary, as well as useful links.

Resolves https://github.com/adoptium/ci-jenkins-pipelines/issues/1091

### CODE REVIEW

**Review by karianna — APPROVED** [review]

File: tools/nightly_build_and_test_stats.groovy [hunk 1]
```diff
@@ -325,6 +359,86 @@ def verifyReleaseContent(String version, String release, String variant, Map sta
...
+                // Then we exit this lambda and skip to the next platform.
+                return
+            }
+
+            def pipelineLink = trssURL+"/api/getAllChildBuilds?parentId="+trssId+"\\&buildNameRegex=^"+jdkVersion+"\\-"+platformConversionMap[onePlatform][0]+"\\-temurin\$"
+            def trssBuildJobNames = sh(returnStdout: true, script: "wget -q -O - ${pipelineLink}")
+            def platformResult = "???% - Build not found. Pipeline link: " + pipelineLink
+
+            // Does this platform have a build in this pipeline?
+            if ( trssBuildJobNames.length() > 10 ) {
```
Comment (andrew-m-leonard) [inline tools/nightly_build_and_test_stats.groovy:388]: Not quite sure of the significance of "10" here?

File: tools/nightly_build_and_test_stats.groovy [hunk 2]
```diff
@@ -325,6 +359,86 @@ def verifyReleaseContent(String version, String release, String variant, Map sta
...
+                buildJobNamesJson.each { buildJob ->
+                    platformResult = "???% - Build found, but no reproducibility tests. Build link: " + buildJob.buildUrl
+                    def testPlatform = platformConversionMap[onePlatform][1]
+                    def reproTestName=platformReproTestMap[onePlatform][1]
+                    def reproTestBucket=platformReproTestMap[onePlatform][0]
+                    def testJobTitle="Test_openjdk${jdkVersionInt}_hs_${reproTestBucket}_${testPlatform}.*"
+                    def trssTestJobNames = sh(returnStdout: true, script: "wget -q -O - ${trssURL}/api/getAllChildBuilds?parentId=${buildJob._id}\\&buildNameRegex=^${testJobTitle}\$")
+
+                    // Did this build have tests?
+                    if ( trssTestJobNames.length() > 10 ) {
```
Comment (andrew-m-leonard) [inline tools/nightly_build_and_test_stats.groovy:401]: ditto "10", if significant maybe add reason in comment?

**Review by andrew-m-leonard — COMMENTED** [review]
Just one comment on the significance of "10" ?

**Comment by andrew-m-leonard** [comment]
I've updated the live Slack job to use this PR, to see how it looks for a few days...

File: tools/nightly_build_and_test_stats.groovy [hunk 3]
```diff
@@ -728,8 +885,13 @@ node('worker') {
                     }
                 }
 
+                def reproducibilityText = "Reproducibility: N/A"
```
Comment (andrew-m-leonard) [inline tools/nightly_build_and_test_stats.groovy:888]: If it's for a version that is not in reproducibleBuilds, then lets not output the "Reproducibility: N/A" text, as jdk8,11,17 will never be there...

**Review by andrew-m-leonard — CHANGES_REQUESTED** [review]

File: tools/nightly_build_and_test_stats.groovy (hunk 1 above)
Comment (adamfarley) [reply tools/nightly_build_and_test_stats.groovy:388]: Hi Andrew. The significance of 10 was me noticing that even if the wget returns an empty array, the square brackets still give us a string length of 2. On the first draft, I didn't know if there was a space in the middle (there isn't) so I just went with 10 on the basis that a returned string with a non-empty json would have dozens of characters.

I've changed this to 2.

File: tools/nightly_build_and_test_stats.groovy (hunk 2 above)
Comment (adamfarley) [reply tools/nightly_build_and_test_stats.groovy]: As above. :)

File: tools/nightly_build_and_test_stats.groovy (hunk 3 above)
Comment (adamfarley) [reply tools/nightly_build_and_test_stats.groovy:888]: Okie dokie.

Comment (adamfarley) [reply tools/nightly_build_and_test_stats.groovy]: Done, and I adjusted the spaces too so the kerning is ok.

**Comment by andrew-m-leonard** [comment]
> @andrew-m-leonard - Thanks for the reviews. I've tweaked the relevant code. Let me know if this works for you. :)

@adamfarley I've just done a run using your latest code: https://adoptium.slack.com/archives/C09NW3L2J/p1728486568438139

The jdk-21.0.5+9-ea has run with tests enabled... here: https://trss.adoptium.net/allTestsInfo?buildId=66ff33dad24e1b006e7b75f0&limit=1&hasChildren=false
so we need to check why it thinks **tests were disabled**....?

**Comment by adamfarley** [comment]
> > @andrew-m-leonard - Thanks for the reviews. I've tweaked the relevant code. Let me know if this works for you. :)
> 
> @adamfarley I've just done a run using your latest code: https://adoptium.slack.com/archives/C09NW3L2J/p1728486568438139
> 
> The jdk-21.0.5+9-ea has run with tests enabled... here: https://trss.adoptium.net/allTestsInfo?buildId=66ff33dad24e1b006e7b75f0&limit=1&hasChildren=false so we need to check why it thinks **tests were disabled**....?

Ok, I think I've fixed it. I'm rerunning the [test job](https://ci.adoptium.net/job/adam_temp_job/) now.

File: tools/nightly_build_and_test_stats.groovy [hunk 4]
```diff
@@ -666,23 +788,59 @@ node('worker') {
                     }
                 } else {
                     // Check if build in-progress
-                    inProgressBuildUrl = getInProgressBuildUrl(trssUrl, variant, featureRelease, status['expectedReleaseName'].replaceAll("-beta", ""), status['upstreamTag']+"_adopt")
+                    (probableBuildUrl, probableBuildIdForTRSS, probableBuildStatus) = getBuildUrl(trssUrl, variant, featureRelease, status['expectedReleaseName'].replaceAll("-beta", ""), status['upstreamTag']+"_adopt")
```
Comment (andrew-m-leonard) [inline tools/nightly_build_and_test_stats.groovy:791]: This "getInProgressBuildUrl" was originally added to purely see if A build was in-progress for the given tag, regardless of what platform, is was just a "rough" status as to "is there anything still going...?"
Thus it's intention was not for finding a "specific" build, which is what I think we need now, the situation can be that in TRSS there could be several entries for a given tag say "jdk-21.0.5+9-ea", which has #334 and #335, as Windows_aarch64 was re-run... As it stands the above will find #335(WindowsAarch64) and incorrectly think no ReproducibleTest ran for x64Linux, which is in #334.

We could update verifyReleaseContent() to add the list of platforms to the status[] json (eg.status['foundAssets']), and update getBuildUrl to take a build platform, and find the TRSS build entry for the given tag that has that build platform...?

Comment (adamfarley) [reply tools/nightly_build_and_test_stats.groovy:791]: After in-person discussion, me and Andrew have agreed that per-platform pipeline identification is the optimum way to produce reproducibility percentages in the long run.

However, in the spirit of getting an MVP merges asap, we will be merging the code as-is, and breaking out the per-platform solution into a separate issue (link shortly).

Comment (adamfarley) [reply tools/nightly_build_and_test_stats.groovy:791]: https://github.com/adoptium/ci-jenkins-pipelines/issues/1132

**Review by andrew-m-leonard — APPROVED** [review]

**Pull request merged**